src/client/client.ts

```diff
@@ -1,235 +1,108 @@
+      this.logger
+    ).unwrap();
+
+    let metricReporter: MetricReporter | undefined;
+    if (!this.config.disableReporting) {
+      metricReporter = new MetricReporter(
+        this.superJson,
+        this.config,
+        this.logger
+      );
+      hookMetrics(this.events, metricReporter);
+      metricReporter.reportEvent({
         eventType: 'SDKInit',
         occurredAt: new Date(),
       });
     }
-    registerFailoverHooks(this.hookContext, this);
-  }
-
-  /**
-   * @deprecated
-   * This is not a part of the public API, DON'T USE THIS METHOD
-   * Returns a BoundProfileProvider that is cached according to `profileConfig` and `providerConfig` cache keys.
-   */
-  async cacheBoundProfileProvider(
-    profileConfig: ProfileConfiguration,
-    providerConfig: ProviderConfiguration
-  ): Promise<BoundProfileProvider> {
-    const cacheKey = profileConfig.cacheKey + providerConfig.cacheKey;
 
-    const bound = this.boundCache[cacheKey];
-
-    const now = Math.floor(Date.now() / 1000);
-    //If we don't have anything in cache we must bind
-    if (bound === undefined) {
-      await this.rebind(profileConfig, providerConfig);
-      //If we do but timeout is expired we schedule rebind
-    } else if (bound.expiresAt < now) {
-      void Promise.resolve().then(() =>
-        this.rebind(profileConfig, providerConfig)
-      );
-    }
+    registerFailoverHooks(this.events, this.logger);
 
-    return this.boundCache[cacheKey].profileProvider;
-  }
-
-  private async rebind(
-    profileConfig: ProfileConfiguration,
-    providerConfig: ProviderConfiguration
-  ): Promise<void> {
-    const cacheKey = profileConfig.cacheKey + providerConfig.cacheKey;
-
-    const now = Math.floor(Date.now() / 1000);
-
-    const profileProvider = new ProfileProvider(
+    this.internal = new InternalClient(
+      this.events,
       this.superJson,
-      profileConfig,
-      providerConfig,
-      this
+      this.config,
+      NodeFileSystem,
+      this.boundProfileProviderCache,
+      this.logger
     );
-
-    const boundProfileProvider = await profileProvider.bind({
-      security: providerConfig.security,
-    });
-    this.boundCache[cacheKey] = {
-      profileProvider: boundProfileProvider,
-      expiresAt: now + Config.instance().superfaceCacheTimeout,
-    };
   }
 
   /** Gets a provider from super.json based on `providerName`. */
   async getProvider(providerName: string): Promise<Provider> {
-    const providerSettings = this.superJson.normalized.providers[providerName];
-
-    if (providerSettings === undefined) {
-      throw unconfiguredProviderError(providerName);
-    }
-
-    return new Provider(
-      this,
-      new ProviderConfiguration(providerName, providerSettings.security)
-    );
+    return getProvider(this.superJson, providerName);
   }
 
   /** Returns a provider configuration for when no provider is passed to untyped `.perform`. */
   async getProviderForProfile(profileId: string): Promise<Provider> {
```

Comment:
nitpick: I can probably `git blame` this all the way back to myself, but maybe we should rename this to be more descriptive. Like `getDefaultProviderForProfile` or something like that 🤷‍♀️

Reply:
I did not go into breaking changes yet